### PR TITLE
Set receive_window per quic connection

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
Differentiate the staked and non-staked connections and set different receive_window. Testing has show that with everything equal, by tweaking the receive window alone, the receive_window is impacting the chunks_received only below about 10 * PACKET_DATA_SIZE. Beyond that -- it is not making much difference. 

#### Summary of Changes
This change sets the receive_window for non-staked node to 1 * PACKET_DATA_SIZE, and maps the staked nodes's connection's receive_window between 1.2 * PACKET_DATA_SIZE to  10 * PACKET_DATA_SIZE based on the stakes.

The changes is based on Quinn library change to support per connection receive_window tweak at the server side. https://github.com/quinn-rs/quinn/pull/1393

Fixes #

### CODE REVIEW

**Comment by lijunwangs** [comment]
[Receive window’s impact to chunks_received..pdf](https://github.com/solana-labs/solana/files/9264402/Receive.window.s.impact.to.chunks_received.pdf)

File: sdk/src/quic.rs [hunk 1]
```diff
@@ -12,3 +12,15 @@ pub const QUIC_KEEP_ALIVE_MS: u64 = 1_000;
 // applications. Different applications vary, but most seem to
 // be in the 30-60 second range
 pub const QUIC_CONNECTION_HANDSHAKE_TIMEOUT_MS: u64 = 60_000;
+
+/// The receive window for QUIC connection from unstaked nodes is
+/// set to this ratio times [`solana_sdk::packet::PACKET_DATA_SIZE`]
+pub const QUIC_UNSTAKED_RECEIVE_WINDOW_RATIO: u64 = 1;
+
+/// The receive window for QUIC connection from minum staked nodes is
```
Comment (pgarg66) [inline sdk/src/quic.rs:20]:
```suggestion
/// The receive window for QUIC connection from minimum staked nodes is
```

File: streamer/src/nonblocking/quic.rs [hunk 2]
```diff
@@ -305,6 +319,51 @@ fn prune_unstaked_connections_and_add_new_connection(
...
+fn compute_recieve_window(
+    max_stake: u64,
+    min_stake: u64,
+    peer_type: ConnectionPeerType,
+    peer_stake: u64,
+) -> VarInt {
+    match peer_type {
+        ConnectionPeerType::Unstaked => {
+            VarInt::from_u64((PACKET_DATA_SIZE as u64 * QUIC_UNSTAKED_RECEIVE_WINDOW_RATIO) as u64)
+                .unwrap()
```
Comment (pgarg66) [inline streamer/src/nonblocking/quic.rs:357]: I think we should handle this unwrap(). With current limits its guaranteed to succeed, but maybe it is not future proof.
How about if this function returns a `Result<VarInt,..>` and the caller sets the receive window only if the result is `Ok`?

**Review by pgarg66 — COMMENTED** [review]
Looks good except one review comment.

File: streamer/src/nonblocking/quic.rs (hunk 2 above)
Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: Done

**Review by ryleung-solana — COMMENTED** [review]
I'm kind of confused about the intent of this... if I'm understanding correctly, this sets the total amount of data the connection can receive. If the sender sends more than that, the server will close the connection. What if the the sender wants to send more (e.g. it has a big batch of transactions to send)? It seems kind of inconvenient to have your connection repeatedly dropped in the middle of a batch send, only to be forced to reconnect.

**Comment by lijunwangs** [comment]
> 



> I'm kind of confused about the intent of this... if I'm understanding correctly, this sets the total amount of data the connection can receive. If the sender sends more than that, the server will close the connection. What if the the sender wants to send more (e.g. it has a big batch of transactions to send)? It seems kind of inconvenient to have your connection repeatedly dropped in the middle of a batch send, only to be forced to reconnect.

receive_window != max_data. receive_window is Quinn's implementation to achieve the flow control described by Quinn's max_data mechanism. As data is read, max_data is credited and increased and communicated back to the client periodically. If the sender is exceeding the data limit, it becomes blocked and signals that with DATA_BLOCKED frame. Quinn implementation will ensure that not to exceed the MAX_DATA from the client side. If the sender maliciously ignore MAX_DATA, the connection will be disconnected by the server. In my tests with bench-tps I have not seen connection being remade.

**Review by pgarg66 — APPROVED** [review]

**Review by sakridge — COMMENTED** [review]
lgtm

File: streamer/Cargo.toml [hunk 3]
```diff
@@ -21,7 +21,9 @@ nix = "0.24.2"
 pem = "1.0.2"
 percentage = "0.1.0"
 pkcs8 = { version = "0.8.0", features = ["alloc"] }
-quinn = "0.8.3"
+quinn = {git = "https://github.com/quinn-rs/quinn.git", branch = "0.8.x", commit = "37c19743cc881cf71369946d572849d5d2ffc3fd"}
+quinn-proto = {git = "https://github.com/quinn-rs/quinn.git", branch = "0.8.x", commit = "37c19743cc881cf71369946d572849d5d2ffc3fd"}
```
Comment (behzadnouri) [inline streamer/Cargo.toml:25]: @lijunwangs This does not seem correct. I am seeing warnings:
```
warning: solana/streamer/Cargo.toml: unused manifest key: dependencies.quinn-proto.commit                                                           
warning: solana/streamer/Cargo.toml: unused manifest key: dependencies.quinn.commit 
```

You need to use `rev` instead:
https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html#cargotoml-vs-cargolock
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories

as in: 
https://github.com/solana-labs/solana/blob/773a4dd4d/Cargo.toml#L100

Comment (lijunwangs) [reply streamer/Cargo.toml:25]: Thanks. @behzadnouri -- will take care of it.

File: core/src/staked_nodes_updater_service.rs [hunk 4]
```diff
@@ -36,11 +36,15 @@ impl StakedNodesUpdaterService {
                     let mut new_ip_to_stake = HashMap::new();
                     let mut new_id_to_stake = HashMap::new();
                     let mut total_stake = 0;
+                    let mut max_stake: u64 = 0;
+                    let mut min_stake: u64 = u64::MAX;
```
Comment (behzadnouri) [inline core/src/staked_nodes_updater_service.rs:40]: This is computing `{min,max}_stake` but never actually updating those fields in `shared_staked_nodes`.
This is patched in https://github.com/solana-labs/solana/pull/31082

Until #31082 is merged, these two fields are always zero, so any benchmarks or tests done so far have been wrong.

Comment (lijunwangs) [reply core/src/staked_nodes_updater_service.rs:40]: Sorry, I do not understand, the min max is updated in try_refresh_stake_maps line 76-82. What am I missing?

Comment (behzadnouri) [reply core/src/staked_nodes_updater_service.rs:40]: The respective fields in `shared_staked_nodes` are not updated:
https://github.com/solana-labs/solana/blob/4564bcdc1/core/src/staked_nodes_updater_service.rs#L54-L57

Comment (behzadnouri) [reply core/src/staked_nodes_updater_service.rs:40]: These fields:
https://github.com/solana-labs/solana/blob/03abaf76d/streamer/src/streamer.rs#L31-L32
are always zero.

Comment (lijunwangs) [reply core/src/staked_nodes_updater_service.rs:40]: Got it. Thanks! The results of attached PDF of comparing different receive_window size was done via hard coded value in the code when I set the receive_window per connection -- as we did not have a mechanism to simulate the stakes in the bench-tps tool.

**Pull request merged**